### PR TITLE
Fix: false negative of `object-shorthand` (fixes #6429)

### LINKED ISSUE #6429: object-shorthand not being enforced when using "avoidQuotes"
**What version of ESLint are you using?**
2.12.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
either
```
"object-shorthand": ["error", "always", {"avoidQuotes": true}],
```
or
```
/*eslint object-shorthand: ["error", "always", { "avoidQuotes": true }]*/
```

**What did you do? Please include the actual source code causing the issue.**
Object shorthand is not being enforced. We configured the rule in our eslintrc and have an object block that should fail. As a minimal test case I have created a repo with a single file and configures the rule inline in the file. That should definitely cause it to fail, but `npm test` passes.

https://github.com/TheSavior/eslint-shorthand-testcase

I'm sure this is user error since the eslint test suite is passing, but we haven't been able to figure out what we are doing wrong with this rule. Other rules are working for us no problem.

### PR DESCRIPTION
**Pull request by mysticatea**
Fixes #6429.

I refactored the logic flow. The new flow has 3 parts.

1. Check whether it should be ignored always or not.
2. Check for concise properties.
3. Check for longform properties.

Then I fixed it for #6429.

### CODE REVIEW

File: lib/rules/object-shorthand.js [hunk 1]
```diff
@@ -129,19 +129,8 @@ module.exports = {
...
-                    context.report(node, "Expected longform " + type + " syntax.");
-                }
-
-                // {'xyz'() {}} should be written as {'xyz': function() {}}
-                if (AVOID_QUOTES && isStringLiteral(node.key) && isConciseProperty) {
-                    context.report(node, "Expected longform method syntax for string literal keys.");
-                }
-
-                // at this point if we're concise or if we're "never" we can leave
-                if (APPLY_NEVER || AVOID_QUOTES || isConciseProperty) {
```
Comment (mysticatea) [inline lib/rules/object-shorthand.js]: The cause was this `AVOID_QUOTES`.
This has been replaced by L166-168 & L177-179.

**Comment by nzakas** [comment]
@mysticatea can you rebase?

**Comment by mysticatea** [comment]
Sure, I rebased.

**Comment by nzakas** [comment]
Lgtm

**Pull request merged**